assert,util: fix constructor lookup in deep equal comparison

BridgeAR opened the pull request:
The latest performance deep equal optimization did not take into account that
an object may have a property called constructor. This is addressed
in this PR by adding a new fast path and using fallbacks.

On top of that, this adds a few performance optimizations for the constructor
name in util.inspect().

ljharb reviewed (approved):
i didn't thoroughly check that the list of constructors and prototypes is exhaustive, but LGTM

jasnell commented on lib/internal/util/comparisons.js:
```diff
@@ -17,18 +35,65 @@ const {
 const { isError } = require('internal/util');
+const { Buffer } = require('buffer');
+
+const wellKnownConstructors = new SafeSet();
```
Can these be added by passing an Array into the `new SafeSet` constructor? Would likely be more efficient to do at once rather than the individual add calls.

BridgeAR replied:
I wanted to do that. We have a primordial test that prevents this.
I am happy to remove that, if we agree on that.

ljharb replied:
Rather than forgo the robustness, SafeSet's constructor could be implemented so that it uses SafeArrayIterator to get a safe iterator on the argument, and then pass that to `super` - then you'd be able to use the array without having to worry about anything, or without the ergonomic nightmare of lots of `.add` calls.

If that's interesting I could throw up a quick PR to do that, and land that first?

BridgeAR replied:
That would definitely be useful. I don't think the order of the PRs matter in this case. Either of us has to adjust the code afterwards.

aduh95 replied:
The reason I never did that was because `SafeArrayIterator` is (was?) not very performant, and forcing the code author to wrap the array explicitly would nudge them into writing it in a more performant way. FWIW, I find chaining the `.add` calls can help make it less wordy:
```js
const wellKnownConstructors = new SafeSet()
  .add(Array)
  .add(ArrayBuffer)
  // ...
  .add(WeakSet);

BridgeAR replied:
That would indeed be nicer, I'll update the code

jasnell commented on lib/internal/util/inspect.js:
```diff
@@ -613,16 +630,26 @@ wellKnownPrototypes.set(ArrayPrototype, { name: 'Array', constructor: Array });
 wellKnownPrototypes.set(ArrayBufferPrototype, { name: 'ArrayBuffer', constructor: ArrayBuffer });
```
Should these include `__proto__: null`?

BridgeAR replied:
We only look up these specific two properties, so I would not know how that could possibly make any difference.

ljharb replied:
it should in theory, to avoid a future footgun, but I agree it won't matter in practice.

BridgeAR replied:
I would keep it as is for now, if that's fine for you. I don't think we benefit from the change here.

BridgeAR commented:
I accidentally already included Float16Array and that ended up with undefined as constructor and wrong behavior. I fixed that now and added an additional test case to also detect different null prototypes properly.

aduh95 commented on lib/internal/util/comparisons.js:
```diff
@@ -1,14 +1,32 @@
+  Date,
   DatePrototypeGetTime,
   Error,
+  Float16Array,
```
https://github.com/nodejs/node/blob/4fd5db479a8b37a48c565a79eeb719a0fb459469/lib/eslint.config_partial.mjs#L359-L363
```suggestion
  globalThis: { Float16Array },
```

aduh95 commented on lib/internal/util/comparisons.js:
```diff
@@ -17,18 +35,65 @@ const {
+wellKnownConstructors.add(DataView);
+wellKnownConstructors.add(Date);
+wellKnownConstructors.add(Error);
+if (Float16Array) { // TODO(BridgeAR): Remove when regularly supported
+  wellKnownConstructors.add(Float16Array);
+}
```
We probably do not need the condition here, it should be harmless to add `undefined` to the set – I'm fine keeping it to be clear

BridgeAR replied:
That would cause null prototypes to not properly being distinguishable anymore. I already added a test case to cover that.
